wlauto/utils/statedetect.py

```diff
@@ -0,0 +1,147 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+"""
+State detection functionality for revent workloads. Uses OpenCV to analyse screenshots from the device.
+Requires a 'statedetection' directory in the workload directory that includes the state definition yaml file,
+and the 'templates' folder with PNGs of all templates mentioned in the yaml file.
+
+Requires the following Python libraries:
+numpy, pyyaml (yaml), imutils and opencv (cv2)
+
+"""
+
+import os
+
+import yaml
+try:
+    import numpy as np
+except ImportError:
+    np = None
+try:
+    import cv2
+except ImportError:
+    cv2 = None
+try:
+    import imutils
+except ImportError:
+    imutils = None
+
+
+class StateDefinitionError(RuntimeError):
+    def __init__(self, arg):
+        self.args = arg
+
+
+def auto_canny(image, sigma=0.33):
+    # compute the median of the single channel pixel intensities
+    v = np.median(image)
+
+    # apply automatic Canny edge detection using the computed median
+    lower = int(max(0, (1.0 - sigma) * v))
+    upper = int(min(255, (1.0 + sigma) * v))
+    edged = cv2.Canny(image, lower, upper)
+
+    # return the edged image
+    return edged
+
+
+def match_state(screenshot_file, defpath):
+    # check dependencies
+    if np == None or cv2 == None or imutils == None:
+        raise RuntimeError("State detection requires numpy, opencv (cv2) and imutils.")
+
+    # load and parse state definition file
+    statedefs_file = os.path.join(defpath, 'definition.yaml')
+    if not os.path.isfile(statedefs_file):
+        raise StateDefinitionError("Missing state definitions yaml file: "+statedefs_file)
+    with open(statedefs_file) as fh:
+        state_definitions = yaml.load(fh)
+
+    # check if file exists, then load screenshot into opencv and create edge map
+    if not os.path.isfile(screenshot_file):
+        raise StateDefinitionError("Screenshot file not found")
+    img_rgb = cv2.imread(screenshot_file)
+    img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)
+    img_edge = auto_canny(img_gray)
+
+    # make a list of all templates defined in the state definitions
+    template_list = []
+    for state in state_definitions["workload_states"]:
+        template_list.extend(state["templates"])
+
+    # check all template PNGs exist
+    missing_files = 0
+    for template_png in template_list:
+        if not os.path.isfile(os.path.join(defpath, 'templates', template_png+'.png')):
+            missing_files += 1
```

Comment:
rather than counting the missing files then raising an exception later why not raise it here, you can even specify which file is missing